### PR TITLE
New: Add rule to check the usage of `theme-color` meta tag

### PR DESCRIPTION
**Pull request by alrra**
Ref: https://html.spec.whatwg.org/multipage/semantics.html#meta-theme-color

### CODE REVIEW

File: packages/rule-meta-theme-color/README.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        <title>example</title>
+        ...
+    </head>
+    <body>...</body>
+</html>
+```
+
+<!-- markdownlint-disable MD038 -->
+
+The `theme-color` meta tag is wrongly specified as ` theme-color`:
```
Comment (molant) [inline packages/rule-meta-theme-color/README.md]: Maybe add a note that there's an space at the beginning to make it clear?

File: packages/rule-meta-theme-color/src/rule.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        },
+        id: 'meta-theme-color',
+        schema: [],
+        scope: RuleScope.any
+    }
+
+    public constructor(context: RuleContext) {
+
+        /*
+         * This function exists because not all connector (e.g.: jsdom)
```
Comment (molant) [inline packages/rule-meta-theme-color/src/rule.ts]: > connectors

File: packages/rule-meta-theme-color/src/rule.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                    !rgbaIsSupported)
+            ) {
+                await context.report(resource, element, `'content' attribute value ('${contentValue}') is not unsupported`);
+            }
+        };
+
+        const validate = async (event: TraverseEnd) => {
+            const { resource }: { resource: string } = event;
+
+            // The following check don't make sense for non-HTML documents.
```
Comment (molant) [inline packages/rule-meta-theme-color/src/rule.ts]: > doesn't

File: packages/rule-meta-theme-color/src/rule.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+            if (!isHTMLDocument(resource, context.pageHeaders)) {
```
Comment (molant) [inline packages/rule-meta-theme-color/src/rule.ts]: Isn't `TraverseEnd` send only for HTML documents? If so this check is not needed.

File: packages/rule-meta-theme-color/src/rule.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            // Check if meta tag is specified.
+
+            if (themeColorMetaTags.length === 0) {
+                await context.report(resource, null, `No 'theme-color' meta tag was specified`);
+
+                return;
+            }
+
+            /*
+             * Treat the first charset meta tag as the one
```
Comment (molant) [inline packages/rule-meta-theme-color/src/rule.ts]: Copy paste from `charset meta tag`?

File: packages/rule-meta-theme-color/src/rule.ts (hunk 4 above)
Comment (molant) [reply packages/rule-meta-theme-color/src/rule.ts]: Any reason why you are not subscribed to `traverse::element::meta`? This way you are sure to only receive the element for HTML, you can remove parts of the code and you only need to keep track if a previous `theme-color` has been emitted already.

You only need to subscribe to `traverse::end` and if none found previously report it.

File: packages/rule-meta-theme-color/src/rule.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+            const bodyMetaTags: Array<IAsyncHTMLElement> = getThemeColorMetaTags(await pageDOM.querySelectorAll('body meta'));
+
+            if ((bodyMetaTags.length > 0) && bodyMetaTags[0].isSame(themeColorMetaTag)) {
+                await context.report(resource, themeColorMetaTag, `Meta tag should not be specified in the '<body>'`);
+
+                return;
+            }
+
+            // All other charset meta tags should not be included.
```
Comment (sarvaje) [inline packages/rule-meta-theme-color/src/rule.ts]: Review this comment

File: packages/rule-meta-theme-color/README.md (hunk 1 above)
Comment (sarvaje) [reply packages/rule-meta-theme-color/README.md]: Yes, please, the fist time I though it didn't support the short version of the color hehehe.

File: packages/rule-meta-theme-color/src/rule.ts (hunk 5 above)
Comment (alrra) [reply packages/rule-meta-theme-color/src/rule.ts]: Yeap.

File: packages/rule-meta-theme-color/src/rule.ts (hunk 4 above)
Comment (alrra) [reply packages/rule-meta-theme-color/src/rule.ts]: Done.

**Comment by alrra** [comment]
@molant @sarvaje Should be ready for another review.

File: packages/rule-meta-theme-color/README.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    ...
+}
+```
+
+## What does the rule check?
+
+The rule checks if a single `theme-color` meta tag is specified in the
+`<head>` and the value of its `content` attribute is a [valid CSS
+color][color] supported by the [targeted
+browsers](https://sonarwhal.com/docs/user-guide/further-configuration/browser-context/).
```
Comment (molant) [inline packages/rule-meta-theme-color/README.md]: Shouldn't this be a foot link?

File: packages/rule-meta-theme-color/src/rule.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    public constructor(context: RuleContext) {
+
+        let bodyElementWasReached: boolean = false;
+        let firstThemeColorMetaTag: IAsyncHTMLElement;
+
+        const checkIfThemeColorMetaTagWasSpecified = async (event: TraverseEnd) => {
+            const { resource } = event;
+
+            /*
+             * TODO: Remove this when `traverse::end` is
```
Comment (molant) [inline packages/rule-meta-theme-color/src/rule.ts]: Did you open an issue about it? We should link to it.

Comment (alrra) [reply packages/rule-meta-theme-color/src/rule.ts]: @molant It's basically the same issue as https://github.com/sonarwhal/sonarwhal/issues/982.

File: packages/rule-meta-theme-color/README.md (hunk 7 above)
Comment (alrra) [reply packages/rule-meta-theme-color/README.md]: Fixed.

File: packages/rule-meta-theme-color/src/rule.ts (hunk 8 above)
Comment (molant) [reply packages/rule-meta-theme-color/src/rule.ts]: Can you link to it in the code?

Comment (alrra) [reply packages/rule-meta-theme-color/src/rule.ts]: Done.

**Pull request merged**